packages/xo-remote-parser/src/index.js

```diff
@@ -2,68 +2,83 @@ import filter from 'lodash/filter'
+    const val = parsed[key]
+    parsed[key] = val === 'true' ? true : val === 'false' ? false : JSON.parse(val)
+  })
+  return parsed
+}
+
+const makeOptionList = options => {
+  const encoded = {}
 
-  const [type, rest] = string.split('://')
+  Object.keys(options).forEach(key => {
+    const val = options[key]
+    encoded[key] = val === 'true' || val === 'false' ? val : JSON.stringify(val)
+  })
+  return queryString.stringify(encoded)
+}
+
+export const parse = string => {
+  let object = {}
+  let [type, rest] = string.split('://')
   if (type === 'file') {
     object.type = 'file'
+    let optionList
+    ;[rest, optionList] = rest.split('?')
     object.path = `/${trimStart(rest, '/')}` // the leading slash has been forgotten on client side first implementation
+    object = { ...parseOptionList(optionList), ...object }
   } else if (type === 'nfs') {
     object.type = 'nfs'
-    let host, port, path
+    let host, port, path, optionList
     // Some users have a remote with a colon in the URL, which breaks the parsing since this commit: https://github.com/vatesfr/xen-orchestra/commit/fb1bf6a1e748b457f2d2b89ba02fa104554c03df
     try {
-      ;[, host, port, path] = NFS_RE.exec(rest)
+      ;[, host, port, path, optionList] = NFS_RE.exec(rest)
     } catch (err) {
       ;[host, path] = rest.split(':')
       object.invalidUrl = true
     }
     object.host = host
     object.port = port
     object.path = `/${trimStart(path, '/')}` // takes care of a missing leading slash coming from previous version format
+    object = { ...parseOptionList(optionList), ...object }
   } else if (type === 'smb') {
     object.type = 'smb'
-    const [, username, password, domain, host, path = ''] = SMB_RE.exec(rest)
+    const [, username, password, domain, host, path = '', optionList] = SMB_RE.exec(rest)
     object.host = host
     object.path = path
     object.domain = domain
     object.username = username
     object.password = password
+    object = { ...parseOptionList(optionList), ...object }
   } else if (type === 's3' || type === 's3+http') {
-    const parsed = urlParser(string, true)
+    const parsed = new URL(string)
     object.protocol = parsed.protocol === 's3:' ? 'https' : 'http'
     object.type = 's3'
     object.region = parsed.hash.length === 0 ? undefined : parsed.hash.slice(1) // remove '#'
     object.host = parsed.host
     object.path = parsed.pathname
     object.username = parsed.username
     object.password = decodeURIComponent(parsed.password)
-    const qs = parsed.query
-    object.allowUnauthorized = qs.allowUnauthorized === 'true'
+    object = { ...parseOptionList(parsed.search), ...object }
   }
   return object
 }
 
-export const format = ({
-  type,
-  host,
-  path,
-  port,
-  username,
-  password,
-  domain,
-  protocol = type,
-  region,
-  allowUnauthorized = false,
-}) => {
+export const format = ({ type, host, path, port, username, password, domain, protocol = type, region, ...options }) => {
```

Comment:
I feel like we should store options in a dedicated property…

Reply:
I don't think it's a good idea to have a dedicated property, especially if we change the storage format later. Also it guarantee that we don't have a property password and a password